e2e/tests/s3/put_object.feature

```diff
@@ -0,0 +1,16 @@
+Feature: PutObject
+
+  Scenario: PutObject: regular file
+    When a file is uploaded to 'data' bucket using 'mydir/myfile' key
+    Then the file is stored as 'data/mydir/myfile'
+
+  Scenario: PutObject: symlink
+    When a file is uploaded to 'data' bucket using 'myfile' key
+     And a symlink is uploaded to 'links' bucket using 'mydir/mylink' key pointing to 'data/myfile'
+    Then the file is accessible via 'links/mydir/mylink'
+
+  Scenario: PubObject: invalid keys: root component
+    Given keys with leading '/' are considered invalid
```

Comment:
@xDarksome that's a bad testing patter by two reasons:
1. You hide test data behind step definition. It's not clear what is tested.
2. It doesn't follow `preset -> assert` pattern. Scenario witn only `Given` step looks weird.

Reply:
Either [Scenario Outline](https://cucumber-rs.github.io/cucumber/current/writing/scenario_outline.html) or [data tables](https://cucumber-rs.github.io/cucumber/current/writing/data_tables.html) should be used here.

Reply:
@xDarksome 

> Does it really matter to the person who reads the scenario? There is always the question of how low/high level those step descriptions should be.

Yes. As such a one person I was wondering quite long looking at thos steps and trying to understand what is going on, until I read the steps definition.

Having explicit test cases in scenarios makes it also better to extend/change them without touching step definitions.

> Didn't find any guidelines related to this.

More of tribal knowledge.

> Making two steps and stroring useless data in the `World` feels even more wrong.

No. That's exactly what `World` is created for.

Reply:
> Yes. As such a one person I was wondering quite long looking at thos steps and trying to understand what is going on, until I read the steps definition.

I presume this wasn't obvious?
```gherkin
Given keys containing '.', '..', '//' path components are considered invalid
```
 The problem is that the "path component" term isn't specific enough in this context?

> More of tribal knowledge.

So which tribe you got it from? :smile: 

---

Whatever,  what scenario structure do you want to see here?

Will the following do?
```gherkin
  Scenario: PubObject: invalid keys: root component
    When trying to upload files with the following keys: 
    | /abc   |
    | /abc/d |
    Then 'InvalidArgument' error is returned

  Scenario: PutObject: invalid keys: current/parent dir and empty components
    When trying to upload files with the following keys: 
    | ./abc    |
    | abc/.    |
    | abc/./d  |
    | ../abc   |
    | abc/..   |
    | abc/../d |
    | abc//    |
    | abc//d   |
    Then 'InvalidArgument' error is returned
```

Btw, we should stop using ` for variable quotation, none of the syntax highlighters that I've seen understand it. Let's use '

Reply:
@tyranron ^

Reply:
@xDarksome 

> Btw, we should stop using ` for variable quotation, none of the syntax highlighters that I've seen understand it.

Have no fuss with this in IDEA. This is regular Markdown quotation. Don't like to diverge much in docs/commets/specs.

> Will the following do?

Yup.

Reply:
@tyranron 

> This is regular Markdown quotation

We are not writing Markdown, it's Gherkin.

Gilab/Guthub/VSCode don't highlight `-quoted Gherkin variables.

See:

```gherkin
Given a = `1`
```

```gherkin
Given a = '1'
```

Reply:
@xDarksome Gherkin [doesn't specify](https://cucumber.io/docs/gherkin/reference/) any variable/parameter notion. It allows arbitrary text as descriptions. So highlighting `'1'` quotation is not the Gherkin thing, but the choice of concrete syntax highlighters.

Reply:
@tyranron 

> but the choice of concrete syntax highlighters.

Still, most of them highlight only '

Do IDEA highlight both ` and ' ?

Reply:
@xDarksome 

> Do IDEA highlight both ` and ' ?

IDEA doesn't both.

Reply:
@xDarksome because it diverges from the docs/comments style. It's easier to just keep everything in Markdown style.